transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,418 @@
+struct QuicMuxerInner<C: Crypto> {
+    /// Accept incoming streams.
+    accept_incoming: bool,
+    /// Endpoint channel.
+    endpoint: ConnectionChannel<C>,
+    /// Inner connection object that yields events.
+    connection: Connection<C::Session>,
+    /// Connection waker.
+    waker: Option<Waker>,
+    /// Connection timer.
+    timer: Option<Timer>,
+    /// State of all open substreams.
+    substreams: HashMap<StreamId, SubstreamState>,
+    /// Pending substreams.
+    pending_substreams: VecDeque<Waker>,
+    /// Close waker.
+    close_waker: Option<Waker>,
+}
+
+/// State of a single substream.
+#[derive(Debug, Default)]
+struct SubstreamState {
+    /// Waker to wake if the substream becomes readable.
+    read_waker: Option<Waker>,
+    /// Waker to wake if the substream becomes writable.
+    write_waker: Option<Waker>,
+}
+
+impl<C: Crypto> QuicMuxer<C> {
+    pub fn new(endpoint: ConnectionChannel<C>, connection: Connection<C::Session>) -> Self {
+        Self {
+            inner: Mutex::new(QuicMuxerInner {
+                accept_incoming: false,
+                endpoint,
+                connection,
+                waker: None,
+                timer: None,
+                substreams: Default::default(),
+                pending_substreams: Default::default(),
+                close_waker: None,
+            }),
+        }
+    }
+
+    pub fn is_handshaking(&self) -> bool {
+        self.inner.lock().connection.is_handshaking()
+    }
+
+    pub fn peer_id(&self) -> Option<PeerId> {
+        let inner = self.inner.lock();
+        let session = inner.connection.crypto_session();
+        C::peer_id(session)
+    }
+
+    pub fn local_addr(&self) -> Multiaddr {
+        let inner = self.inner.lock();
+        let ip = inner
+            .connection
+            .local_ip()
+            .unwrap_or(IpAddr::V4(Ipv4Addr::UNSPECIFIED));
+        let addr = SocketAddr::new(ip, inner.endpoint.port());
+        crate::transport::socketaddr_to_multiaddr(&addr)
+    }
+
+    pub fn remote_addr(&self) -> Multiaddr {
+        let inner = self.inner.lock();
+        let addr = inner.connection.remote_address();
+        crate::transport::socketaddr_to_multiaddr(&addr)
+    }
+
+    pub(crate) fn set_accept_incoming(&self, accept: bool) {
+        let mut inner = self.inner.lock();
+        inner.accept_incoming = accept;
+    }
+}
+
+impl<C: Crypto> StreamMuxer for QuicMuxer<C> {
+    type Substream = StreamId;
+    type OutboundSubstream = ();
+    type Error = QuicMuxerError;
+
+    fn poll_event(
+        &self,
+        cx: &mut Context,
+    ) -> Poll<Result<StreamMuxerEvent<Self::Substream>, Self::Error>> {
+        let mut inner = self.inner.lock();
+        let now = Instant::now();
+
+        while let Poll::Ready(event) = inner.endpoint.poll_channel_events(cx) {
+            inner.connection.handle_event(event);
+        }
+
+        let _max_datagrams = inner.endpoint.max_datagrams();
```

Comment:
Does this method call have any side effects?

Reply:
no, this is when using quinn-proto from git which changes the api

Reply:
https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/connection/mod.rs#L411